builder/proxmox/step_finalize_template_config.go

```diff
@@ -56,6 +56,28 @@ func (s *stepFinalizeTemplateConfig) Run(ctx context.Context, state multistep.St
 		changes["ide2"] = "none,media=cdrom"
 	}
 
+	if c.CloudInit {
+		vmParams, err := client.GetVmConfig(vmRef)
+		if err != nil {
+			err := fmt.Errorf("Error fetching template config: %s", err)
+			state.Put("error", err)
+			ui.Error(err.Error())
+			return multistep.ActionHalt
+		}
+
+		cloudInitStoragePool := c.CloudInitStoragePool
+		if cloudInitStoragePool == "" {
+			if vmParams["bootdisk"] != nil && vmParams[vmParams["bootdisk"].(string)] != nil {
+				bootDisk := vmParams[vmParams["bootdisk"].(string)].(string)
+				cloudInitStoragePool = strings.Split(bootDisk, ":")[0]
+			}
+		}
+		if cloudInitStoragePool != "" {
+			ui.Say("Adding a cloud-init cdrom in storage pool " + cloudInitStoragePool)
+			changes["ide2"] = cloudInitStoragePool + ":cloudinit"
```

Comment:
Sorry, apparently missed this in the first reading - does this have to be `ide2`? This way, it mixes a bit with the `UnmountISO` option since they both use the same device.
Not sure if this is an actual use case, just a reflection (someone could be wanting the original ISO to be kept mounted, and also use cloudinit?)

Reply:
Yes, that's right. I took ide2 because in all examples they use ide2 for the cloudinit drive.
As far as i see, https://github.com/Telmate/proxmox-api-go uses ide2 hardcoded when you pass an iso file to CreateVm().
You can have ide0..ide3, so i will leave ide2 untouched and look for an unused one. (i think, it's best to start from ide3, so that the cloudinit drive is far away from the adapters people will intentionally use for cdroms later. And i will also look at terraform, i think it makes sense to have an option there which removes the cloudinit drive after boot/initi at all)

Reply:
```
                if cloudInitStoragePool != "" {
                        ideControllers := []string{"ide3", "ide2", "ide1", "ide0"}
                        cloudInitAttached := false
                        // find a free ide controller
                        for _, controller := range ideControllers {
                                if vmParams[controller] == nil {
                                        ui.Say("Adding a cloud-init cdrom in storage pool " + cloudInitStoragePool)
                                        changes[controller] = cloudInitStoragePool + ":cloudinit"
                                        cloudInitAttached = true
                                        break
                                }
                        }
                        if cloudInitAttached == false {
                                err := fmt.Errorf("Found no free ide controller for a cloud-init cdrom")
                                state.Put("error", err)
                                ui.Error(err.Error())
                                return multistep.ActionHalt
                        }
                }
```
I added also a test case with 4 iso files. (Why anybody would build a template with several cdroms is another question. "Inserting" such media should be left to terraform or vagrant imho)

Reply:
Indeed a good question, but nice fix! :ok_hand: